platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/impl/InstanceKeyImpl.java

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2015 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.machine.server.impl;
+
+import org.eclipse.che.api.machine.server.spi.InstanceKey;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Map based implementation of {@link org.eclipse.che.api.machine.server.spi.InstanceKey}.
+ *
+ * @author Sergii Kabashniuk
+ */
+
+public class InstanceKeyImpl implements InstanceKey {
+
+    private final Map<String, String> fields;
+
+    public InstanceKeyImpl() {
+        fields = new HashMap<>();
+    }
+
+    public InstanceKeyImpl(Map<String, String> fields) {
+        this.fields = new HashMap<>(fields);
+    }
+
+    public InstanceKeyImpl(InstanceKey instanceKey) {
+        this.fields = new HashMap<>(instanceKey.getFields());
+    }
+
+    /**
+     * @return unmodifiable copy of fields.
+     */
+    @Override
+    public Map<String, String> getFields() {
+        return Collections.unmodifiableMap(fields);
```

Comment:
It is not possible to change field after instance creation. Why you create unmodifiable map each time this method is called? I would prefer use it once.